server/src/main/java/io/unitycatalog/server/persist/DeltaCommitRepository.java

```diff
@@ -109,18 +170,206 @@ private static void handleNormalCommit(
               "Commit version must be the next version after the latest commit %d, but got %d",
               lastCommitVersion, newCommitVersion));
     }
-    // TODO: check commit limit after implementing backfill
+    // getLatestBackfilledVersion may or may not be null because normal commit may or may
+    // not notify a backfill in the same request
+    Optional<Long> latestBackfilledVersion =
+        Optional.ofNullable(commit.getLatestBackfilledVersion());
+    if (latestBackfilledVersion.filter(x -> x > lastCommitVersion).isPresent()) {
+      throw new BaseException(
+          ErrorCode.INVALID_ARGUMENT,
+          String.format(
+              "Latest backfilled version %d cannot be greater than the last commit version = %d",
+              latestBackfilledVersion.get(), lastCommitVersion));
+    }
+    long effectiveBackfilledVersion =
```

Comment:
this variable is only used for checking commit limit. cant this be pushed into the checkCommitLimit function? seems like a better abstraction.